src/libraries/System.Private.CoreLib/src/System/Collections/HashHelpers.cs

```diff
@@ -101,10 +101,9 @@ public static uint FastMod(uint value, uint divisor, ulong multiplier)
             // which allows to avoid the long multiplication if the divisor is less than 2**31.
             Debug.Assert(divisor <= int.MaxValue);
 
-            ulong lowbits = multiplier * value;
-            // 64bit * 64bit => 128bit isn't currently supported by Math https://github.com/dotnet/runtime/issues/31184
-            // otherwise we'd want this to be (uint)Math.BigMul(lowbits, divisor, out _)
-            uint highbits = (uint)((((lowbits >> 32) + 1) * divisor) >> 32);
+            // This is equivalent of (uint)Math.BigMul(multiplier * value, divisor, out _). This version
+            // is faster than BigMul currently because we only need the high bits.
```

Comment:
This could use `(uint)Bmi2.X64.MultiplyNoFlags(multiplier * value, divisor)` directly as that overload is fast (has no memory access overhead).

Reply:
Issue with that is it pushes Dictionary out of R2R compilation

Reply:
That might be worth a comment too (but don't reset CI for it 🙂